tower-util/src/ready.rs

```diff
@@ -8,27 +8,114 @@ use std::{
 };
 use tower_service::Service;
 
-/// Future yielding a `Service` once the service is ready to process a request
+/// A future that yields the service when it is ready to accept a request.
 ///
-/// `Ready` values are produced by `ServiceExt::ready`.
-pub struct Ready<'a, T, Request> {
-    inner: &'a mut T,
+/// `ReadyOneshot` values are produced by `ServiceExt::ready_oneshot`.
+pub struct ReadyOneshot<T, Request> {
+    inner: Option<T>,
     _p: PhantomData<fn() -> Request>,
 }
 
 // Safety: This is safe because `Services`'s are always `Unpin`.
+impl<T, Request> Unpin for ReadyOneshot<T, Request> {}
+
+impl<T, Request> ReadyOneshot<T, Request>
+where
+    T: Service<Request>,
+{
+    #[allow(missing_docs)]
+    pub fn new(service: T) -> Self {
+        Self {
+            inner: Some(service),
+            _p: PhantomData,
+        }
+    }
+}
+
+impl<T, Request> Future for ReadyOneshot<T, Request>
+where
+    T: Service<Request>,
+{
+    type Output = Result<T, T::Error>;
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        ready!(self
+            .inner
+            .as_mut()
+            .expect("poll after Poll::Ready")
+            .poll_ready(cx))?;
+
+        Poll::Ready(Ok(self.inner.take().expect("poll after Poll::Ready")))
+    }
+}
+
+impl<T, Request> fmt::Debug for ReadyOneshot<T, Request>
+where
+    T: fmt::Debug,
+{
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_struct("ReadyOneshot")
+            .field("inner", &self.inner)
+            .finish()
+    }
+}
+
+/// A future that yields a mutable reference to the service when it is ready to accept a request.
+///
+/// `ReadyAnd` values are produced by `ServiceExt::ready_and`.
+pub struct ReadyAnd<'a, T, Request>(ReadyOneshot<&'a mut T, Request>);
```

Comment:
Would `ReadyMut` make sense here?

Reply:
Really I want this to be called just `Ready`, but for the time being that's not really possible. In the interrim though, I find `ReadyMut` a weird name (I know tokio has used it in the past) because it's not really important that it holds something that is `mut`. It's more important that it holds something that is `&` (as opposed to owned), but `ReadyRef` would also be kind of weird (since it _doesn't_ contain "mut"). I went with `ReadyAnd` because it has a nice cadence to it: "await this service being ready and ...". It also then matches the method name on `ServiceExt`, which I think reads better as `ready_and` than as `ready_mut`.

Reply:
@jonhoo why don't we do a breaking change and bump version number? We could also consider combining crates to make this process easier.

Reply:
I figured we might want to do that eventually, but didn't want to block this on a major version bump. I'm guessing there are some other changes we'd want to make for that too, so that would likely take longer?

Reply:
Yeah, lets just ship this as is then and we can do the nice stuff later.